contracts/staking/GoodFundManager.sol

```diff
@@ -189,76 +237,127 @@ contract GoodFundManager is DAOContract {
-		// );
-        address reserveAddress = nameService.addresses(nameService.RESERVE());
-		lastTransferred = block.number.div(blockInterval);
-		ERC20 iToken = ERC20(_staking.iToken());
-		// iToken balance of the reserve contract
-		uint256 currentBalance = iToken.balanceOf(reserveAddress);
-		// collects the interest from the staking contract and transfer it directly to the reserve contract
-		//`collectUBIInterest` returns (iTokengains, tokengains, precission loss, donation ratio)
-	    _staking.collectUBIInterest(reserveAddress);
-
-		// Finds the actual transferred iToken
-		uint256 interest =
-			iToken.balanceOf(reserveAddress).sub(currentBalance);
-		//uint256 effectiveInterest =
-		//	interest.mul(avgEffectiveStakedRatio).div(DECIMAL1e18);
-		//uint256 interestDonated = interest.sub(effectiveInterest);
+    /**
+     * @dev Collects UBI interest in iToken from a given staking contract and transfers
+     * that interest to the reserve contract. Then transfers the given gd which
+     * received from the reserve contract back to the staking contract and to the
+     * bridge, which locks the funds and then the GD tokens are been minted to the
+     * given address on the sidechain
+     */
+    function collectInterest(address[] memory _stakingContracts)
+        public
+        reserveHasInitialized
+        //requireDAOContract(address(_staking))
+    {
+        // require(
+        //     canRun(),
+        //     "Need to wait for the next interval"
+        // );
+        uint initialGas = gasleft();
+        lastTransferred = block.number.div(blockInterval);
+        ERC20 iToken = ERC20(nameService.getAddress("CDAI"));
+        // iToken balance of the reserve contract
+        uint256 currentBalance = iToken.balanceOf(nameService.addresses(nameService.RESERVE()));
+        
+        for(uint i = _stakingContracts.length - 1; i >= 0; i--){ // zxelements are sorted by balances from lowest to highest 
+		
+            if(_stakingContracts[i] != address(0x0)){
+                StakingContract(_stakingContracts[i]).collectUBIInterest(
+                    nameService.addresses(nameService.RESERVE()));
+                    }
+                   
+            if(i == 0) break; // when active contracts length is 1 then gives error
+        }
+       
+        // Finds the actual transferred iToken
+        uint interest = iToken.balanceOf(nameService.addresses(nameService.RESERVE())) - currentBalance;
+        uint gasPriceIncDAI = getGasPriceInCDAI(initialGas);
+        if (block.timestamp >= lastCollectedInterest + 5184000){ // 5184000 is 2 months in seconds
+            require(interest >= gasPriceIncDAI, "Collected interest should be bigger than spent gas"); // This require is necessary to keeper can not abuse this function
+        }else{
+            require(interest >= 4 * gasPriceIncDAI, "Collected interests should be at least 4 times bigger than gas cost since last call of this function sooner than 2 months");
+        }
```

Comment:
this calculation and require should be the very last thing.
see how keeper network doing it, see how we do gas reimbursement in AdminWallet.
it should be calculated over actual gas used not initial gas.